### PR TITLE
Add custom task tutorial

### LINKED ISSUE #1011: [FEAT][CV] Custom Task Tutorial
Create a tutorial showing how to run deepchecks on a task that is not detection or classification.

Use the pedestrian dataset https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html as a segmentation task.

Also: in this tutorial, demonstrate defining custom property (label / prediction) and using it in the appropriate checks to get insights about the segmentation problem.

### PR DESCRIPTION
**Pull request by matanper**
#### Reference Issues/PRs

resolve #1011

### CODE REVIEW

File: docs/source/tutorials/vision/custom_task_tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,280 @@
+====================
+Custom Task Tutorial
+====================
+
+Computer vision is consists of different objectives models are trained for. The desired objective reflects
```
Comment (noamzbr) [inline docs/source/tutorials/vision/custom_task_tutorial.rst:5]:
```suggestion
Computer vision is an umbrella term for a wide spectrum of objectives models are trained for. These objective reflect
```

File: docs/source/tutorials/vision/custom_task_tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+on the structure of the data and the possible actions on it.
+
+In Deepchecks the first step before running checks is to create an implementation of `VisionData` which
```
Comment (noamzbr) [inline docs/source/tutorials/vision/custom_task_tutorial.rst:8]:
```suggestion
The first step before running any Deepchecks checks is to create an implementation of `VisionData`, which
```

File: docs/source/tutorials/vision/custom_task_tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+standardize the structure of the data for a given computer vision task.
```
Comment (noamzbr) [inline docs/source/tutorials/vision/custom_task_tutorial.rst:9]: link to tutorials

File: docs/source/tutorials/vision/custom_task_tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+Some of the checks are agnostic to the specific task and will work on all of them, and some are tied to the
```
Comment (noamzbr) [inline docs/source/tutorials/vision/custom_task_tutorial.rst:11]: This part should be improved - start by saying what current tasks are supported, then give example for checks that can run for other tasks and explain why is that (the last part is what you explained here).

File: docs/source/tutorials/vision/custom_task_tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+specific details of a task and won't work for other tasks. When implementing your own task all of the
+task-agnostic checks will be able to run.
+
+Defining the Data & Model
+=========================
+
+First we will define a `Dataset` of COCO-128 segmentation task.
```
Comment (noamzbr) [inline docs/source/tutorials/vision/custom_task_tutorial.rst:18]: maybe note that this is a setup stage and that the deepchecks specific code will be in the next section?

File: docs/source/tutorials/vision/custom_task_tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,280 @@
+====================
+Custom Task Tutorial
```
Comment (noamzbr) [inline docs/source/tutorials/vision/custom_task_tutorial.rst:2]: pretty long, needs table of contents?

File: docs/source/tutorials/vision/custom_task_tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+      """Class for loading the COCO segmentation dataset."""
+
+      def get_classes(self, batch_labels) -> List[List[int]]:
+          """Return per label a list of classes (by id) in it."""
+          # The input `batch_labels` is the result of `batch_to_labels` function.
+          return batch_labels[0]
+
+      def batch_to_labels(self, batch):
+          """Extract from the batch only the labels."""
+          return batch[1], batch[2]
```
Comment (noamzbr) [inline docs/source/tutorials/vision/custom_task_tutorial.rst:229]: explain what happens here

File: docs/source/tutorials/vision/custom_task_tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+      def infer_on_batch(self, batch, model, device):
+          """Infer on a batch of images."""
+          predictions = model.to(device)(batch[0])
+          return predictions
+
+      def batch_to_images(self, batch) -> Iterable[np.ndarray]:
+          """Convert the batch to a list of images as (H, W, C) 3D numpy array per image."""
+          return [tensor.numpy().transpose((1, 2, 0)) for tensor in batch[0]]
+
+Now we are able to run checks that are working only on the image data, since it's in the standard format of
```
Comment (noamzbr) [inline docs/source/tutorials/vision/custom_task_tutorial.rst:240]:
```suggestion
Now we are able to run checks that use only the image data, since it's in the standard Deepchecks format
```
Link to explanation of image format.

File: docs/source/tutorials/vision/custom_task_tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+Deepchecks. For more checks we'll need to define custom properties or metrics.
```
Comment (noamzbr) [inline docs/source/tutorials/vision/custom_task_tutorial.rst:241]:
```suggestion
Deepchecks. In order for additional check types to run, we'll need to define custom properties or metrics.
```

File: docs/source/tutorials/vision/custom_task_tutorial.rst [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+
+In order to run checks that are using label or prediction properties we we'll have to implement
+a custom property. For a more in-depth explanation on properties see
+:doc:`Data Properties </user-guide/vision/vision_properties>`
+
+.. code-block:: python
+
+  #### Label properties
+
+  def number_of_detections(labels) -> List[int]:
```
Comment (noamzbr) [inline docs/source/tutorials/vision/custom_task_tutorial.rst:254]: would have been cooler to use area

File: docs/source/tutorials/vision/custom_task_tutorial.rst [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+  label_properties = [
+    {'name': '# Detections per Label', 'method': number_of_detections, 'output_type': 'discrete'},
+    {'name': 'Classes in Labels', 'method': classes_in_labels, 'output_type': 'class_id'}
+  ]
+
+
+Implement Custom Metric
+=======================
+
+Some checks requires a metric in order to run. When using a custom task you will also have to create a custom
```
Comment (noamzbr) [inline docs/source/tutorials/vision/custom_task_tutorial.rst:276]:
```suggestion
Some checks test the model performance and requires a metric in order to run. When using a custom task you will also have to create a custom
```

File: docs/source/tutorials/vision/custom_task_tutorial.rst [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+metric in order for those checks to work, since the built-in metrics does not know to handle your data structure.
```
Comment (noamzbr) [inline docs/source/tutorials/vision/custom_task_tutorial.rst:277]:
```suggestion
metric in order for those checks to work, since the built-in metrics don't know to handle your data structure.
```

**Review by noamzbr — CHANGES_REQUESTED** [review]
Main improvements that I think we must do:

1. Add an example of actually using checks (e.g. Property Drift & Single Feature Contribution). It would make the result complete and give a better sense of what can I "get" if I have a custom task. 
2. Migrate this example to a .py file, so this code will be testable and will produce an output that can be part of the docs.

File: docs/source/tutorials/vision/custom_task_tutorial.rst (hunk 10 above)
Comment (matanper) [reply docs/source/tutorials/vision/custom_task_tutorial.rst:254]: Maybe, but it's harder and not the focus of the tutorial

File: deepchecks/vision/checks/methodology/simple_feature_contribution.py [hunk 13]
```diff
@@ -120,8 +116,8 @@ def update(self, context: Context, batch: Any, dataset_kind: DatasetKind):
 
             properties['target'] += classes
         else:
-            raise DeepchecksValueError(
-                f'Check {self.__class__.__name__} does not support task type {dataset.task_type}')
+            imgs = batch.images
+            properties['target'] += dataset.get_classes(batch.labels)
```
Comment (noamzbr) [inline deepchecks/vision/checks/methodology/simple_feature_contribution.py:120]: maybe just use get_classes in all of them and be agnostic?

**Review by noamzbr — APPROVED** [review]

File: deepchecks/vision/checks/methodology/simple_feature_contribution.py (hunk 13 above)
Comment (matanper) [reply deepchecks/vision/checks/methodology/simple_feature_contribution.py:120]: but in object detection we do extra things on the label anyway (like cropping image and bboxes)

Comment (matanper) [reply deepchecks/vision/checks/methodology/simple_feature_contribution.py]: Done
